tox_pdm/plugin4.py

```diff
@@ -0,0 +1,232 @@
+        clone_pdm_files(self.env_dir, self.core["toxinidir"])
+        self.execute(
+            ["pdm", "use", "-f", self.base_python.extra["executable"]], StdinSource.OFF
+        )
+
+    def _get_python(self, base_python: t.List[str]) -> t.Optional[PythonInfo]:
+        interpreter = next(
+            filter(None, (get_interpreter(p, []) for p in base_python)), None
+        )
+        if not interpreter:
+            return None
+        return PythonInfo(
+            implementation=interpreter.implementation,
+            version_info=interpreter.version_info,
+            version=interpreter.version,
+            is_64=(interpreter.architecture == 64),
+            platform=interpreter.platform,
+            extra={"executable": Path(interpreter.system_executable)},
+        )
+
+    def python_cache(self) -> t.Dict[str, t.Any]:
+        base = super().python_cache()
+        base.update({"executable": str(self.base_python.extra["executable"])})
+        return base
+
+    def env_bin_dir(self) -> Path:
+        return self.package_path / ("Scripts" if os.name == "nt" else "bin")
+
+    def env_python(self) -> Path:
+        return t.cast(Path, self.base_python.extra["executable"])
+
+    def env_site_package_dir(self) -> Path:
+        return self.package_path / "lib"
+
+    def prepend_env_var_path(self) -> t.List[Path]:
+        return [self.env_bin_dir]
+
+
+class PdmRunner(Pdm, PythonRun):
+    def _setup_env(self) -> None:
+        super()._setup_env()
+        groups = self.conf["groups"]
+        cmd = ["pdm", "install", "--no-self"]
+        for group in groups:
+            cmd.extend(("--group", group))
+        self.execute(cmd, StdinSource.OFF)
+        bin_path = self.env_site_package_dir() / "bin"
+        if not bin_path.exists():
+            return
+        scripts_path = self.env_bin_dir()
+        for file in bin_path.iterdir():
+            shutil.move(os.fspath(file), os.fspath(scripts_path))
+
+    def register_config(self) -> None:
+        super().register_config()
+        self.conf.add_config(
+            "groups",
+            of_type=t.List[str],
+            default=[],
+            desc="Specify the dependency groups to install",
+        )
+
+    @property
+    def _default_package_tox_env_type(self) -> str:
+        return "pdm-pep-517"
+
+    @staticmethod
+    def id() -> str:
+        return "pdm"
+
+
+class PdmPackageEnv(Pdm, PackageToxEnv):
+    def register_config(self) -> None:
+        super().register_config()
+        self.conf.add_config(
+            keys=["pkg_dir"],
+            of_type=Path,
+            default=lambda conf, name: self.env_dir / "dist",
+            desc="directory where to put project packages",
+        )
+
+    @property
+    def pkg_dir(self) -> Path:
+        return t.cast(Path, self.conf["pkg_dir"])
+
+    def perform_packaging(self, for_env: EnvConfigSet) -> t.List[Package]:
+        of_type: str = for_env["package"]
+        cmd = [
```

Comment:
Does this create a PEP-517 env all the time? Or does PDM under the hod caches the isolated build env?